### PR TITLE
twister: fix stuck issue after flash time out

### PR DESCRIPTION
**Pull request by ghost**
We may have more than one sub-process when flash, so kill the whole
flash process group instead of single process to avoid twister stuck.

Signed-off-by: Guo Lixin <lixinx.guo@intel.com>

### CODE REVIEW

**Review by PerMac — CHANGES_REQUESTED** [review]
Does this code causes any issues? It looks exactly like an example from `communicate()` docs https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate. After `TimeoutExpired` exception is caught, the process is manually killed and `communicate()`  is called there again to collect stdout and stderr that could get left in pipes.
So I don't understand the need to change it. Won't we loose stderr, stdout that could still be left in pipes without reading it with communicate after killing the process?

**Comment by LixinGuoX** [comment]
> Does this code causes any issues? It looks exactly like an example from `communicate()` docs https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate. After `TimeoutExpired` exception is caught, the process is manually killed and `communicate()` is called there again to collect stdout and stderr that could get left in pipes. So I don't understand the need to change it. Won't we loose stderr, stdout that could still be left in pipes without reading it with communicate after killing the process?

Yes, I understand your concern, when we run twister on intel_adsp platform, we may have more than one subprocess, one is west flash, others are for cavstool_client.py, like below:
```
guo      4011310  0.7  0.2  57016 36380 pts/0    S+   09:15   0:00 /usr/bin/python3 /home/guo/.local/bin/west flash --skip-rebuild -d /home/guo/zephyrproject/zephyr/twister-out/intel_adsp_cavs18/tests/ztest/mock/testing.ztest.mock --runner intel_adsp --pty --remote-host=10.239.144.246
guo      4011315  0.0  0.0  26828 13392 pts/0    S+   09:15   0:00 python3 /home/guo/zephyrproject/zephyr/soc/xtensa/intel_adsp/tools/cavstool_client.py -s 10.239.144.246 /home/guo/zephyrproject/zephyr/twister-out/intel_adsp_cavs18/tests/ztest/mock/testing.ztest.mock/zephyr/zephyr.ri.5f0539901d3600c3a76889cabb019654
```

When we have `cavstool_client.py` issues like #47652 , twister will only kill the sub-process for west flash, `cavstool_client.py` will  still stuck, and we will not get any response from `communicate()`, we may want to kill the process group instead of single process for west flash like: `os.killpg(os.getpid(proc.pid), signal.SIGTERM)`

**Comment by chen-png** [comment]
you invoked another python or shell script during the subprocess.Popen of west-flash or device-serial-pty, this will generate a new child process and the return proc is the parent process, when you invoked proc.kill, it will only kill the parent process and leave a defunct child process, and for your platform, it can't support two two child processes at the same time, so it was blocked.

**Review by PerMac — CHANGES_REQUESTED** [review]
os.killpg and os.setsid are UNIX features and are not available on Windows

**Comment by PerMac** [comment]
Maybe you can try with psutils? I played with the second answer from https://stackoverflow.com/questions/4789837/how-to-terminate-a-python-subprocess-launched-with-shell-true and found it working on Windows

**Comment by LixinGuoX** [comment]
> Maybe you can try with psutils? I played with the second answer from https://stackoverflow.com/questions/4789837/how-to-terminate-a-python-subprocess-launched-with-shell-true and found it working on Windows

Yes, that helps a lot, modified as suggested, please re-review, thanks!

**Review by PerMac — DISMISSED** [review]
LGTM

**Review by enjiamai — DISMISSED** [review]

**Review by hakehuang — DISMISSED** [review]

**Review by chen-png — DISMISSED** [review]

**Review by nashif — CHANGES_REQUESTED** [review]
we are adding yet another method for killing a process. There is terminate() in the handler class and another special one in the BinaryHandler and finally one more way to kill a process in the QemuHanlder. Lets try and cleanup and unify this into one single function to deal with all scenarios in a clean way.

File: scripts/pylib/twister/twisterlib/handlers.py [hunk 1]
```diff
@@ -90,13 +90,15 @@ def record(self, harness):
...
+    @staticmethod
+    def terminate(proc_pid):
         # encapsulate terminate functionality so we do it consistently where ever
         # we might want to terminate the proc.  We need try_kill_process_by_pid
         # because of both how newer ninja (1.6.0 or greater) and .NET / renode
         # work.  Newer ninja's don't seem to pass SIGTERM down to the children
         # so we need to use try_kill_process_by_pid.
-        for child in psutil.Process(proc.pid).children(recursive=True):
+        proc = psutil.Process(proc_pid)
+        for child in proc.children(recursive=True):
```
Comment (chen-png) [inline scripts/pylib/twister/twisterlib/handlers.py:101]: what you changed, actually it is the same as the old terminate function, so I guess what you need is to replace all process.kill() or process.terminate() by this self.terminate(process)

Comment (LixinGuoX) [reply scripts/pylib/twister/twisterlib/handlers.py:101]: Yes, changed this function a little to fit for other handlers.

**Comment by LixinGuoX** [comment]
Seems like we have some CI issue with m2gl025_miv and hifive1

**Pull request closed without merging**